### PR TITLE
Initial contribution of the validator

### PR DESCRIPTION
**Pull request by powdercloud**
Written by @gregable and @powdercloud of Google,
with credits given to original versions for the
css parser/tokenizer and the html parser in the file headers.

This is the validator which is already used by the amphtml developer mode;
see also https://github.com/ampproject/amphtml#the-amp-validator

README.md describes how to build this; there's currently only a minimal
ad-hoc build script, build.py, which should work for Ubuntu 14.

We still need to contribute our tests (we do have tests), so this is
only a first step.

### CODE REVIEW

**Comment by cramforce** [comment]
@erwinmombay Could you also take a look.

File: validator/LICENSE [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (cramforce) [inline validator/LICENSE]: No need to repeat here.

File: validator/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<!---
```
Comment (cramforce) [inline validator/CONTRIBUTING.md]: Probably OK to not have this file.

**Comment by cramforce** [comment]
Does it make sense to also check in the generated JSON, so we can use the validator without the complicated build process?

**Comment by erwinmombay** [comment]
@cramforce will do

**Comment by powdercloud** [comment]
I'd like to not check this in initially, just so that there is a clean checkin. I think it's better to figure out what you need to have checked in once it's clear how you want to use it.

To see what this generates (actually Javascript, not json), I stuck it up as well.
https://github.com/powdercloud/amphtml/blob/codegen/validator/codegen/validator-generated.js

**Comment by cramforce** [comment]
Definitely, lets get this in quickly. I'm fine with merging with the 2 comments above resolved.

File: validator/package.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "amp_validator",
+  "version": "0.1.0",
+  "description": "Validator for AMP HTML (www.ampproject.org)",
+  "main": "index.js",
```
Comment (erwinmombay) [inline validator/package.json]: is there actually an index.js file?

**Comment by erwinmombay** [comment]
@cramforce should we whitelist validator/**/*.* for presubmit? (i already assume we need to skip it for lint checks)

**Comment by cramforce** [comment]
@erwinmombay whitelisting everything but the license check is fine (or at least make sure we pass that check). When we make this the master repo for the validator we can bring everything up to pass.

File: validator/testdata/minimum_valid_amp.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the license.
+-->
+<head>
+  <meta charset="utf-8">
+  <link rel="canonical" href="./regular-html-version.html" />
+  <meta name="viewport" content="width=device-width,minimum-scale=1">
+  <style>body {opacity: 0}</style><noscript><style>body {opacity: 1}</style></noscript>
+  <script src="https://ssl.gstatic.com/amphtml/v0.js" async></script>
```
Comment (erwinmombay) [inline validator/testdata/minimum_valid_amp.html]: move `async` as the first prop for best practices. (i can make this change later, just marking it for now)

**Comment by powdercloud** [comment]
@cramforce removed the two files.

**Comment by erwinmombay** [comment]
@cramforce created https://github.com/ampproject/amphtml/pull/953

File: validator/package.json (hunk 3 above)
Comment (powdercloud) [inline validator/package.json]: There is one, but not included in this PR. The purpose of package.json is to let build.py fetch the dependencies, so index.js isn't used for anything, and build.py contains a snippet which generates its own main method, available (when built) as ./codegen/validate. I'll remove this line and see how it goes.

Comment (powdercloud) [inline validator/package.json]: Turns out this "main": entry does nothing, so I got rid of it.

File: validator/testdata/minimum_valid_amp.html (hunk 4 above)
Comment (powdercloud) [inline validator/testdata/minimum_valid_amp.html]: fixed

**Comment by powdercloud** [comment]
Looks like the checks are green now, thanks to you.
If there's something else that I need to be doing to this pull request, please lemme know.

**Comment by erwinmombay** [comment]
LGTM, deferring to @cramforce for final LGTM

**Comment by cramforce** [comment]
LGTM. Merging.

**Pull request merged**